src/rust/engine/src/scheduler.rs

```diff
@@ -331,6 +336,23 @@ impl Scheduler {
     })
   }
 
+  async fn wait_for_tail_tasks(tasks: Vec<BoxFuture<'static, ()>>) {
+    if !tasks.is_empty() {
```

Comment:
You could invert this and early return if `is_empty()`. Will make the function easier to understand w/ less nesting